adr/ADR-11.md

```diff
@@ -0,0 +1,92 @@
+# JetStream API unsigned 64 bit numerical value fields
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-07-15|
+|Author  |@scottf|
+|Status  |Approved|
+|Tags    |server, client, jetstream|
+
+## Context
+
+This document summarizes fields in the JetStream api schema that are unsigned 64 bit numerical values. 
+The most up-to-date information can be found in the [schema](https://github.com/nats-io/jsm.go/tree/main/schemas/jetstream/api/v1)
+
+## Schema Object Fields
+
+Object and list of fields within the object that are unsigned 64 bit values.
+
+#### `consumer_configuration`
+ 
+* `opt_start_seq`
+
+#### `consumer_create_request`
+
+* `opt_start_seq`
+
+#### `consumer_create_response`
+
+* `consumer_seq`
+* `stream_seq`
+
+#### `consumer_info_response`
+
+* `opt_start_seq`
+* `consumer_seq`
+* `stream_seq`
+
+#### `consumer_list_response`
+
+* `opt_start_seq`
+* `consumer_seq`
+* `stream_seq`
+
+#### `stream_configuration`
+ 
+* `opt_start_seq`
+
+#### `stream_create_request`
+
+* `opt_start_seq`
+
+#### `stream_create_response`
+
+* `opt_start_seq`
+
+#### `stream_info_response`
+
+* `opt_start_seq`
+* `first_seq`
+* `last_seq`
+
+#### `stream_purge_response`
+
+* `purged`
+
+
+#### `pub_ack_response`
+
+* `seq`
+
+#### `stream_purge_request`
+
+* `seq`
+* `keep`
+
+## Message Meta Data
+
+The message meta data carried in the reply_to field of a JetStream message contains meta data. 
+All numerical fields are unsigned 64 bit values.
```

Comment:
The real issue which is not really addressed here (here we are duplicating all the specifications) - is that if we are using JSON (which we are), then the max sequence number cannot exceed `9007199254740991`. This is within the positive range of numbers in JSON, which are only `2^53 - 1` - this means that we should be thinking on how to address this "limitation" by having a generational ID for the sequence, and the number to compute an effective sequence number regardless of the programming language used to access these values.

As it is custom parsing in the case of JavaScript is only possible if the value were expressed as a string. In all cases, we shouldn't duplicate the schema documentation (that is what the schema is).

Reply:
1. JSON doesn't care, it just defines a number as digits. Maybe Javascript cares, or schema tools care. Regardless, the server is already coded to use uint64. Even if it was changed to use int64, 2^63 -1 is still larger than 2^53 -1, which means almost every piece of numerical data in the system is at risk according to this. But Go, Java and .NET are handling these values fine.

Do you have a solution? 

2. I understand and agree that we should not duplicate data and the schema is the source of truth. If you all would prefer I will move all this to the issue I created to make sure that client parity is achieved, but I did edit the ADR to reference the schema as the most up to date information, and it doesn't hurt to call this issue out, as demonstrated by the conversation we are having.

Reply:
>I understand and agree that we should not duplicate data and the schema is the source of truth. If you all would prefer I will move all this to the issue I created to make sure that client parity is achieved, but I did edit the ADR to reference the schema as the most up to date information, and it doesn't hurt to call this issue out, as demonstrated by the conversation we are having.

I think the conversation, thrashing out design etc is valuable and the ADR should focus on calling out what is needed in a broad sense and perhaps present some trategies for handling the large numbers.  Calling out individual fields - all of them the same - doesn't seem that useful and ensure that this ADR will be out of date soon. We might also propose that the server cap sequences to a number lower than uint64 while still storing the data in uint64 if we find supporting the full 64bit space is too problematic.

If we focussed on the problem as a whole and pull in call outs to the various other data types that also have to be handled correctly - for example specific int32's, languages should not send bigger numbers than int32 - that would make this ADR useful.

Combined with updating the source material (schemas) it would leave us in a really good place and this ADR being valuable

Reply:
> ) JSON doesn't care, it just defines a number as digits. Maybe Javascript cares, or schema tools care. Regardless, the server is already coded to use uint64. Even if it was changed to use int64, 2^63 -1 is still larger than 2^53 -1, which means almost every piece of numerical data in the system is at risk according to this. But Go, Java and .NET are handling these values fine.

I think the IETF would disagree with you. JSON is a data exchange format, if you are not following the spec, you are doing something that looks like JSON but it is not JSON. In fields like headers, the sequence is not JSON, but on the info documents it is. Unfortunately just because one language can represent numbers unsigned in 64-bit space doesn't mean that the documents produce will be properly handled or parsed.

https://datatracker.ietf.org/doc/html/rfc7159#section-6

Reply:
It does not require that. It just says its interoperable:
>Note that when such software is used, numbers that are integers and
   are in the range [-(2**53)+1, (2**53)-1] are interoperable in the
   sense that implementations will agree exactly on their numeric
   values.
   
Which if we care about then the server must change. So again, the question is how should the server behave?

2^53 -1 (9,007,199,254,740,991) seems a sufficiently large number for a sequence (and will fit in a int64, making it easy for clients,) but there will have to be server code to manage what happens when the value gets to ...992